### ISSUE TITLE
Generic Product Field Traversals

### ISSUE THREAD

**Issue by dwincort** [kcsongor/generic-lens#145]
I find it a little frustrating that when I have a sum-of-products data type, I can't easily access the fields by their name.  I do understand the technical limitations here, but I've still been thinking about workarounds.  One that would suit my purposes pretty well is to have a _traversal_ to a field name.

## Motivation and Explanation

Ideally, if I have a data type such as:
```haskell
data Foo
  = Bar { a :: Int }
  | Baz { a :: Int, b :: Bool }
  | Qux { a :: Int, b :: Bool, c :: Char }
  deriving (Generic)
```
and I have a value `Baz 3 True`, that I could do a prism-constructor access of `Baz` followed by a lens-field access on `b`.  Unfortunately, the prism for `Baz` has type `Prism' Foo (Int, Bool)` -- Haskell's lack of anonymous records strikes again.

At a user level, I can make data types for all of these, as in:
```haskell
data Foo = Bar Bar' | Baz Baz' | Qux Qux' deriving Generic
data Bar' = Bar' { a :: Int }
...
```
Not only is this some annoying type bloat, but this creates the problem that I no longer have the free lens `field @"a" :: Lens' Foo Int`.

I could also define, for instance,
```haskell
bazB :: Prism' Foo Bool
quxB :: Prism' Foo Bool
quxC :: Prism' Foo Char
```
which is a valid, if verbose, option.

As it turns out, there are many cases where one doesn't need the full power of a prism but still wants access to fields that are not defined in every constructor.  Obviously, the library maintainer understands this given the traversals in [Data.Generics.Product.Types](https://hackage.haskell.org/package/generic-lens-2.2.1.0/docs/Data-Generics-Product-Types.html) and [Data.Generics.Product.Param](https://hackage.haskell.org/package/generic-lens-2.2.1.0/docs/Data-Generics-Product-Param.html).  I propose this be extended so that one can get a traversal from just a field name.  I imagine something like:
```haskell
class HasFields (field :: Symbol) s t a b where
  fields :: Traversal s t a b
```
This would be very similar to `HasField` except that it would construct a traversal rather than a lens, and it would not check to make sure that the field is present in every constructor.  I think, as good form, it should probably make sure that the field is present in _at least one_ constructor, but technically, this wouldn't even be _necessary_.

## For Your Consideration

- Is this a good idea?
- Is there an alternative I should consider?
- How much work do you think this is?
- Would you welcome a PR if I made one?

**Comment by kcsongor** [comment]
I think this is a good idea! There's no nice alternative that I'm aware of that achieves this goal in a satisfactory way.

A lot of the lens derivation code can be piggybacked off of, so implementation-wise I don't expect this to require too much work. I would most certainly welcome a PR!

**Comment by adamgundry** [comment]
I agree that this makes a lot of sense to offer. FWIW `optics` has a version of this as [`gafield`](https://hackage.haskell.org/package/optics-core-0.4.1/docs/Optics-Generic.html#v:gafield).

**Comment by googleson78** [comment]
An alternative that I would like to see, but am entirely capable of judging whether it's possible:

Can we not leverage some machinery to make it so that the type returned by using a ctor prism actually does have the fields that it "should"?

The ones that come to mind are:
* Use some extensible records implementation and return an extensible record type from the ctor prism instead of returning a vanilla Haskell tuple. This seems very heavyweight, but at the same time, it sounds like it ought to be doable.
* Write custom `GenericLensTupleN` types and then write instances for them which (somehow??) have the `Fields` instances which are implied by the fields of the used ctor prism, returning a `GenericLensTupleN` instead of a vanilla Haskell tuple. This one is entirely cloudy to me, to the point where I question whether it's even possible without incidentally implementing an extensible records library.

**Comment by Elvecent** [comment]
This would be really nice to have. `lens` generates these traversals through Template Haskell, and having used it in practice, I'd say it's indeed reasonable and convenient. Don't think I have time to implement, though :crying_cat_face:
